src/sentry/api/endpoints/organization_traces.py

```diff
@@ -0,0 +1,160 @@
+from sentry.api.api_owners import ApiOwner
+from sentry.api.api_publish_status import ApiPublishStatus
+from sentry.api.base import region_silo_endpoint
+from sentry.api.bases import NoProjects, OrganizationEventsV2EndpointBase
+from sentry.api.paginator import GenericOffsetPaginator
+from sentry.api.utils import handle_query_errors
+from sentry.models.organization import Organization
+from sentry.search.events.builder import SpansIndexedQueryBuilder
+from sentry.search.events.types import ParamsType, QueryBuilderConfig
+from sentry.snuba.dataset import Dataset
+from sentry.snuba.referrer import Referrer
+
+
+class TraceResult(TypedDict):
+    trace: str
+    numSpans: int
+    name: str | None
+    duration: int
+    spans: list[Mapping[str, Any]]
+
+
+class OrganizationTracesSerializer(serializers.Serializer):
+    field = serializers.ListField(required=True, allow_empty=False, child=serializers.CharField())
+    query = serializers.CharField(required=False)
+    maxSpansPerTrace = serializers.IntegerField(default=1, min_value=1, max_value=100)
+
+
+@region_silo_endpoint
+class OrganizationTracesEndpoint(OrganizationEventsV2EndpointBase):
+    publish_status = {
+        "GET": ApiPublishStatus.EXPERIMENTAL,
+    }
+    owner = ApiOwner.PERFORMANCE
+
+    def get(self, request: Request, organization: Organization) -> Response:
+        try:
+            snuba_params, params = self.get_snuba_dataclass(request, organization)
+        except NoProjects:
+            return Response(status=404)
+
+        serializer = OrganizationTracesSerializer(data=request.GET)
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=400)
+        serialized = serializer.validated_data
+
+        per_page = self.get_per_page(request)
+        limit = per_page * serialized["maxSpansPerTrace"]
+
+        # Abort early if number of trace x spans per trace
+        # exceeds 10_000 (the snuba query limit),
+        if limit > 10_000:
+            raise ParseError(detail="Too many spans per trace.")
+
+        def data_fn(offset: int, limit: int):
+            with handle_query_errors():
+                builder = SpansIndexedQueryBuilder(
+                    Dataset.SpansIndexed,
+                    cast(ParamsType, params),
+                    snuba_params=snuba_params,
+                    query=serialized.get("query", ""),
+                    selected_columns=["trace", "timestamp"] + serialized["field"],
+                    # The orderby is intentionally `None` here as this query is much faster
+                    # if we let Clickhouse decide which order to return the results in.
+                    # This also means we cannot order by any columns or paginate.
+                    orderby=None,
+                    limit=limit,
+                    offset=0,
+                    limitby=("trace", serialized["maxSpansPerTrace"]),
+                    sample_rate=options.get("traces.sample-list.sample-rate"),
```

Comment:
How much does this help? Do full granules get skipped or just rows for results?

Reply:
it limits the amount of data scanned so for large orgs it helps a quite a bit. but we can also test this which is why i made it an option